### DISCUSSION TITLE
Dynamically access ServerConfig parameters

### DISCUSSION THREAD

**Discussion by proj-sashido** [netreconlab/Parse-Swift#57] [Q&A] [answered]
**Terminology:**
* `ServerConfig` = the container of all parameters that is set/managed on the Parse Server
* `AppConfig` = the subclass of ParseConfig in my app
---

My understanding is that if a new `ServerConfig` variable is created, I’d have to hard-code this new variable into my `AppConfig` (which would also require pushing an app update to users) before it can be read and used.

On the Obj-C Parse SDK, I could check if there was an object for a particular key, using `object(forKey:)`. This meant that if and when that new variable was created on `ServerConfig`, my app would already know how to respond to it, and would handle it without the need for an app update to users. 

Is there any way to access new ServerConfig parameters without having to hard code them as variables into my app?

**Comment by cbaker6** [comment]
This is all up to how you design your app on the client side. If your Config adds key/values to a known return type, key/value pairs can be added dynamically. Your app should be designed in a way to handle these dynamic key/pairs. This is one of the reasons the SDK uses protocols and [POP](https://www.pluralsight.com/guides/protocol-oriented-programming-in-swift) (not OOP or classes/subclasses). Note, I’ve supplied simple examples to show that you can basically return anything that is `Deocodable`. Your return type doesn’t have to be just a String, Int, Dictionary, etc. It can be a custom object you made or a type erased value such as `AnyCodable`.

I linked to more examples with objects here: https://github.com/netreconlab/Parse-Swift/discussions/56#discussioncomment-4952126

**Comment by proj-sashido** [comment]
Apologies, I think I might have been ambiguous with my question!
Allow me to clarify a little, with an example. 
This is all purely academic, as I know there’s ways around the problem, but am just curious.

Say I have two apps, called “MyCoolApp” and “CoolerApp"
Say my `ServerConfig` looks like:
```
Type: Object, Parameter: “AppColors”, Value: {“bg”: “0xFFFFFF”, “accent”: “0x00FF00”}
Type: Object, Parameter: “AppColors_MyCoolApp”, Value: {“bg”: “0xA41FFF”, “accent”: “0x00FFFF”}
```

On the previous SDK, I’d have done something like:
```
var appName = AppInfo.Name
guard let config = PFConfig.current() else { return }
if let colors = config.object(forKey: “AppColors_\(appName)”) as? Dictionary<String, Any> {
    self.applyTheme(using: colors)
} else { 
    let colors = config.object(forKey: “AppColors”) as! Dictionary<String, Any>
    self.applyTheme(using: colors)
}
```
and this way, until `AppColors_CoolerApp` exists, it would apply the default back to `AppColors`. But if at some point that parameter _is_ created, then it will apply the new colors.

To my knowledge, I can’t do something similar in Parse-Swift; I’d have to manually enter in each variable into my `AppConfig` struct (regardless of whether it is created on `ServerConfig` or not), or restructure my `AppColors` parameter in `ServerConfig` to be more like:
```
Type: Object, Parameter: “AppColors”, Value: {“default: {“bg”: “0xFFFFFF”, “accent”: “0x00FF00”}, “MyCoolApp”: {“bg”: “0xA41FFF”, “accent”: “0x00FFFF”}}
```
So it can search through the `AppColors` variable to see if “CoolerApp” colors exist.


Is my understanding correct?

**Reply by cbaker6** [reply]
Yes, your understanding is correct and you will have to do what you mentioned:

> To my knowledge, I can’t do something similar in Parse-Swift; I’d have to manually enter in each variable into my AppConfig struct (regardless of whether it is created on ServerConfig or not), or restructure my AppColors parameter in ServerConfig to be more like:

I can look into the possibility of adding something that could feel similar to the way used it in the objective-c SDK, but this is currently low priority as I'm getting the Swift SDK ready for a major release. Also note that the README states: 

> The ParseSwift SDK is not a port of the [Parse-SDK-iOS-OSX SDK](https://github.com/parse-community/Parse-SDK-iOS-OSX) and though some of it may feel familiar, it is not backwards compatible and is designed using [protocol oriented programming (POP) and value types](https://www.pluralsight.com/guides/protocol-oriented-programming-in-swift) instead of OOP and reference types.

So the Swift SDK does not attempt replicate the exact behavior of the objective-c SDK.

**Reply by proj-sashido** [reply]
Yep totally understand, wasn’t trying to criticise the architecture of Parse-Swift; just helping my own brain understand the new SDK better as I try to adopt it 😊

**Comment by cbaker6** [comment, marked as answer]
This feature has been added in #68 and released in version [5.1.0](https://github.com/netreconlab/Parse-Swift/releases/tag/5.1.0)

**Reply by proj-sashido** [reply]
Thanks! this is amazing :)

**Reply by proj-sashido** [reply]
I think I’m struggling a little to use this. 

I tried fetching it as per the approach in #68 as follows:
```swift
// Fetching Config
let fetched: [String: AnyCodable] = try? await ParseConfigCodable.fetch()
```
it gives an error: `Generic parameter 'V' could not be inferred`

Similarly, if I try to create a struct `MyConfig: ParseConfigCodable {}` it gives the error:
```swift
Reference to generic type 'ParseConfigCodable' requires arguments in <…>
```


Could you provide a bit more guidance please?

**Reply by cbaker6** [reply]
Did you add the AnyCodable framework via SPM mentioned in the PR that added this feature before attempting to use the feature? https://github.com/Flight-School/AnyCodable. There are also docs, you can’t make a struct from another struct, https://swiftpackageindex.com/netreconlab/parse-swift/5.4.1/documentation/parseswift/parseconfigcodable 

There are a number of test cases that all pass for this feature and the rest of the features in the SDK. I recommend you start with the test cases when you don’t see a demonstration in the [Playgrounds](https://github.com/netreconlab/Parse-Swift/tree/main/ParseSwift.playground/Pages): https://github.com/netreconlab/Parse-Swift/blob/26b69e33b6c5a1682f32d6a5dea3a9f75db53dbc/Tests/ParseSwiftTests/ParseConfigCodableTests.swift#L246-L259

Also, the code you provided can't compile, as you used `try?`, but your type isn't optional, the code should be one of the following:

```swift
// Fetching Config
let fetched: [String: AnyCodable] = try await ParseConfigCodable.fetch()
let fetched2: [String: AnyCodable]? = try? await ParseConfigCodable.fetch()
```
